### PR TITLE
[flang][OpenMP] Make OpenMP clause representation language-agnostic

### PR DESCRIPTION
**Pull request by kparzysz**
The clause templates defined in ClauseT.h were originally based on flang's parse tree nodes. Since those representations are going to be reused for clang (together with the clause splitting code), it makes sense to separate them from flang, and instead have them based on the actual OpenMP spec (v5.2).

The member names in the templates follow the naming presented in the spec, and the representation (e.g. members) is derived from the clause definitions as described in the spec.

Since the representations of some clauses has changed (while preserving the information), the current code using the clauses (especially the code converting parser::OmpClause to omp::Clause) needs to be adjusted.

This patch does not make any functional changes.

### CODE REVIEW

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 1]
```diff
@@ -31,57 +31,33 @@ static void checkMapType(mlir::Location location, mlir::Type type) {
...
-      return mlir::omp::ScheduleModifier::none;
-    }
-
-    return translateScheduleModifier(modType1);
-  }
+  using Schedule = omp::clause::Schedule;
+  const auto &modifier =
+      std::get<std::optional<Schedule::OrderingModifier>>(clause.t);
+  if (modifier)
+    return translateScheduleModifier(*modifier);
```
Comment (tblah) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:50]: Why don't we need to skip simd stuff any more here?

Comment (kparzysz) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:50]: The simd modifier was a member of the same modifier type as all the other modifiers (i.e. member of the same enum).  In the spec, "simd" is classified as a "chunk=modifier", and in the "agnostic" representation it's a member of the "ChunkModifier" type.  This function gets an OrderingModifier value, which no longer contains simd.

The onus is now on the code creating the clauses to segregate the modifiers properly.

**Review by tblah — APPROVED** [review]
Thank you for explaining. LGTM

**Comment by kparzysz** [comment]
@skatrak @kiranchandramohan I've merged it since it's NFC, and Tom has reviewed it, plus I have more work based on it.  Still, feel free to leave post-commit review comments if you want.

**Review by skatrak — COMMENTED** [review]
Thank you Krzysztof, I have a couple of nits that can be addressed as another PR.

**Pull request merged**